File: arm.64/langs/swift/src/bind.swift
import Foundation

func bind(_ binaryPath: String) -> Cabi? {
    let filename = (binaryPath as NSString).lastPathComponent
    let skips = ["c", "cpp", "rust", "go", "swift", "haskell", "zig", "v", "slint_sample"]
    if skips.contains(filename) { return nil }

    guard let handle = dlopen(binaryPath, RTLD_LAZY | RTLD_LOCAL) else { return nil }

    let moreSym = dlsym(handle, "More")
    let pumpSym = dlsym(handle, "Pump")
    let lessSym = dlsym(handle, "Less")

    guard let pumpPtr = pumpSym else {
        dlclose(handle)
        return nil
    }

    let moreFn = moreSym.map { unsafeBitCast($0, to: MoreFn.self) }
    let pumpFn = unsafeBitCast(pumpPtr, to: PumpFn.self)
    let lessFn = lessSym.map { unsafeBitCast($0, to: LessFn.self) }

    return Cabi(name: filename, path: binaryPath, handle: handle, more: moreFn, pump: pumpFn, less: lessFn)
}

File: arm.64/langs/cpp/src/bind.cpp
#include "type.h"
#include <dlfcn.h>
#include <filesystem>
#include <algorithm>

namespace fs = std::filesystem;

Cabi* bind(const std::string& binary_path) {
    std::string filename = fs::path(binary_path).filename().string();
    std::vector<std::string> skips = {"c", "cpp", "rust", "go", "swift", "haskell", "zig", "v", "slint_sample"};
    if (std::find(skips.begin(), skips.end(), filename) != skips.end()) {
        return nullptr;
    }

    void* handle = dlopen(binary_path.c_str(), RTLD_LAZY | RTLD_LOCAL);
    if (!handle) return nullptr;

    auto more = reinterpret_cast<MoreFn>(dlsym(handle, "More"));
    auto pump = reinterpret_cast<PumpFn>(dlsym(handle, "Pump"));
    auto less = reinterpret_cast<LessFn>(dlsym(handle, "Less"));

    if (!pump) {
        dlclose(handle);
        return nullptr;
    }

    auto* cabi = new Cabi();
    cabi->name = filename;
    cabi->path = binary_path;
    cabi->handle = handle;
    cabi->More = more;
    cabi->Pump = pump;
    cabi->Less = less;
    return cabi;
}
